### ISSUE TITLE
Callback "task exit" methods can be called before all task loop items have returned

### ISSUE THREAD

**Issue by dmsimard** [ansible/ansible#24207]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
Ansible

##### ANSIBLE VERSION
```
ansible 2.3.0.0
```

##### CONFIGURATION
Not relevant

##### OS / ENVIRONMENT
Not relevant

##### SUMMARY
When executing a task that has a list of items (ex: ``with_items``), the "task exit" method can be called inside a callback before the last item has done it's "exit" method.
In a typical callback that only prints to stdout, this means that things would print out of order.

In more involved callbacks, this means there might be data missing (that is expected to be there) when the task ends.

##### STEPS TO REPRODUCE
Use the following callback (drop it in a directory and export ``ANSIBLE_CALLBACK_PLUGINS``):

```python
from __future__ import (absolute_import, division, print_function)

from ansible.plugins.callback import CallbackBase

class CallbackModule(CallbackBase):
    CALLBACK_VERSION = 2.0
    CALLBACK_TYPE = 'notification'
    CALLBACK_NAME = 'callback_race'

    def __init__(self):
        super(CallbackModule, self).__init__()
        self.loop_items = []

    def v2_playbook_on_task_start(self, task, is_conditional, is_handler=False):
        print("playbook_on_task_start")

    def v2_runner_item_on_ok(self, result):
        print("runner_item_on_ok")
        self.loop_items.append(result)

    def v2_runner_on_ok(self, result, **kwargs):
        print("runner_on_ok")
        print(self.loop_items)
```

With the following playbook:
```yaml
- name: Create fake hosts
  hosts: localhost
  gather_facts: no
  tasks:
    - name: Add fake hosts in inventory
      add_host:
        name: "{{ item }}"
        ansible_host: "127.0.0.1"
        ansible_connection: "local"
      with_items:
        - host1
        - host2
        - host3
```
##### EXPECTED RESULTS
All items must always exit ( ``v2_runner_item_on_ok`` ) before the task exits ( ``v2_runner_on_ok`` ).

Example correct behavior:
```
$ansible-playbook playbook.yml 

PLAY [Create fake hosts] *****************************************************************************************************************************************************************************************************************************************************************************************************

TASK [Add fake hosts in inventory] *******************************************************************************************************************************************************************************************************************************************************************************************
playbook_on_task_start
changed: [localhost] => (item=host1)
runner_item_on_ok
changed: [localhost] => (item=host2)
runner_item_on_ok
changed: [localhost] => (item=host3)
runner_item_on_ok
runner_on_ok
[<ansible.executor.task_result.TaskResult object at 0x7f7d639b2650>, <ansible.executor.task_result.TaskResult object at 0x7f7d63a8c1d0>, <ansible.executor.task_result.TaskResult object at 0x7f7d639d6ad0>]

PLAY RECAP *******************************************************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```

##### ACTUAL RESULTS
The behavior is inconsistent and reliably reproduceable.
The task can often exit before the last item has exit.

Example problematic behavior:
```
$ ansible-playbook playbook.yml 

PLAY [Create fake hosts] *****************************************************************************************************************************************************************************************************************************************************************************************************

TASK [Add fake hosts in inventory] *******************************************************************************************************************************************************************************************************************************************************************************************
playbook_on_task_start
changed: [localhost] => (item=host1)
runner_item_on_ok
changed: [localhost] => (item=host2)
runner_item_on_ok
runner_on_ok
[<ansible.executor.task_result.TaskResult object at 0x7febe5456650>, <ansible.executor.task_result.TaskResult object at 0x7febe55301d0>]
changed: [localhost] => (item=host3)
runner_item_on_ok

PLAY RECAP *******************************************************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0   
```

**Comment by bcoca** [comment]
it does not get executed before, the problem is with forking and aync queues, the callback events might get received out of order.

**Comment by bcoca** [comment]
cc @jimi-c

**Comment by emonty** [comment]
FWIW, I actually get very consistent results of the final runner_on_ok always immediately before the final runner_item_on_ok. So much so that I currently store the runner_on_ok result as an attribute on the callback class if I'm in the middle of an items, and emit the results for it after the next item and so far it's always working out properly.

So - it certainly might be async and forking and whatnot, but it shows remarkably consistent behavior.

**Comment by bcoca** [comment]
@emonty that should be true most of the time, but we have had many reports of this happening. I suspect that it is mostly systems with other loads that cause preemption of the python processes/threads and end creating the 'unordering'.

**Comment by emonty** [comment]
\o/ yay async!

@bcoca is it possible anywhere to find out the number of items the task expects?

**Comment by bcoca** [comment]
@emonty not sure what you mean

**Comment by bcoca** [comment]
fixed via #28111
